kyo-core/shared/src/main/scala/kyo/StreamCoreExtensions.scala

```diff
@@ -622,6 +620,89 @@ object StreamCoreExtensions:
         ): Stream[V2, Abort[E] & Async & S & S2] =
             mapChunkParUnordered(Async.defaultConcurrency, defaultAsyncStreamBufferSize)(f)(using t1, t2, t3, i1, i2, ev, frame)
 
+        /** Collects values that are emitted by the original stream within the duration [[maxTime]] up to the amount [[maxSize]] and emits
+          * them as a chunk.
+          *
+          * If no elements are emitted by the original stream within [[maxTime]], as soon as any other elements are emitted the result
+          * stream will emit them as a group.
+          *
+          * @param maxSize
+          *   Maximum number of elements to be collected within a single duration. Values of less than one are ignored and treated as one.
+          * @param maxTime
+          *   Maximum amount of time to collect and emit elements
+          * @return
+          *   A new stream that emits collected chunks of elements
+          */
+        def groupedWithin(maxSize: Int, maxTime: Duration, bufferSize: Int = defaultAsyncStreamBufferSize)(using
+            t1: Tag[Emit[Chunk[V]]],
+            t2: Tag[Emit[Chunk[Chunk[V]]]],
+            i1: Isolate.Contextual[S, IO],
+            i2: Isolate.Contextual[S, Abort[E] & Async],
+            ct: SafeClassTag[Closed | E],
+            fr: Frame
+        ): Stream[Chunk[V], S & Abort[E] & Async] =
+            import Event.*
+            enum Event derives CanEqual:
+                case Data(chunk: Chunk[V])
+                case Tick, Flush
+            end Event
+
+            Stream[Chunk[V], S & Abort[E] & Async]:
+                IO.Unsafe {
+                    val safeMax = 1 max maxSize
+                    val channel = Channel.Unsafe.init[Event](1 max bufferSize).safe
+
+                    // Handle loop collecting emitted values and flushing them until completion
+                    val push: Fiber[E | Closed, Unit] < (IO & S) =
+                        Async.run[E | Closed, Unit, S]:
+                            IO.ensure(Async.run[Closed, Unit, Any](channel.put(Flush))):
+                                ArrowEffect.handleLoop(t1, (), stream.emit)(
```

Comment:
You can just get rid of the empty state `()` and the `done = (_, _) => ()` parameter